packages/swingset-liveslots/src/virtualObjectManager.js

```diff
@@ -680,21 +675,53 @@ export function makeVirtualObjectManager(
     }
     // beyond this point, we use 'multifaceted' to switch modes
 
-    if (isDurable) {
-      checkAndUpdateFacetiousness(
-        tag,
-        durableKindDescriptor,
-        facetNames,
-        saveDurableKindDescriptor,
-      );
-    }
+    // The 'stateShape' pattern constrains the `state` of each
+    // instance: which properties it may have, and what their values
+    // are allowed to be. For durable Kinds, the stateShape is
+    // serialized and recorded in the durableKindDescriptor, so future
+    // incarnations (which redefine the kind when they call
+    // defineDurableKind again)
 
     harden(stateShape);
     stateShape === undefined ||
       passStyleOf(stateShape) === 'copyRecord' ||
       Fail`A stateShape must be a copyRecord: ${q(stateShape)}`;
     assertPattern(stateShape);
 
+    if (isDurable) {
+      // durableKindDescriptor is created by makeKindHandle, with just
+      // { kindID, tag, nextInstanceID }, then the first
+      // defineDurableKind (maybe us!) will populate
+      // .facets/.unfaceted and a .stateShape . We'll only see those
+      // properties if we're in a non-initial incarnation.
+
+      assert(durableKindDescriptor);
+
+      // initial creation will update the descriptor with .facets or
+      // .unfaceted, subsequent re-definitions will just assert
+      // compatibility
+      checkAndUpdateFacetiousness(tag, durableKindDescriptor, facetNames);
+
+      const stateShapeCapData = serialize(stateShape);
+
+      // Durable kinds can only hold durable objects in their state,
+      // so if the stateShape were to require a non-durable object,
+      // nothing could ever match. So we require the shape have only
+      // durable objects
+      insistDurableCapdata(vrm, 'stateShape', stateShapeCapData, false);
+
+      // compare against slots of previous definition, incref/decref
+      let oldStateShapeSlots = [];
+      if (durableKindDescriptor.stateShapeCapData) {
+        oldStateShapeSlots = durableKindDescriptor.stateShapeCapData.slots;
+      }
+      const newStateShapeSlots = stateShapeCapData.slots;
+      vrm.updateReferenceCounts(oldStateShapeSlots, newStateShapeSlots);
+      durableKindDescriptor.stateShapeCapData = stateShapeCapData; // replace
```

Comment:
I'm wondering if we should optimize the case where we don't have (and didn't have) a stateShape by omitting it from the descriptor.

Reply:
We could, and I did that for virtual collections' `keyShape`/`valueShape`, but Kinds are low-cardinality, so we're not saving very much space.